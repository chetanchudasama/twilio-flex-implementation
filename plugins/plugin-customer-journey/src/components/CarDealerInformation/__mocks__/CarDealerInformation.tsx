import React from "react";
import { Props } from "../CarDealerInformation";

const CarDealerInformation: React.FC<Props> = (props) => <div />;

export default CarDealerInformation;
