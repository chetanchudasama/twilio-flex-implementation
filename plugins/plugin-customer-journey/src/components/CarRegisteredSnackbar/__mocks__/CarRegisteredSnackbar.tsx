import React from "react";
import { CarRegisteredSnackbarProps } from "../CarRegisteredSnackbar.Props";

export const CarRegisteredSnackbar: React.FC<CarRegisteredSnackbarProps> = (
  props
) => {
  return <div className="car-registered-snackbar">CarRegisteredSnackbar</div>;
};
