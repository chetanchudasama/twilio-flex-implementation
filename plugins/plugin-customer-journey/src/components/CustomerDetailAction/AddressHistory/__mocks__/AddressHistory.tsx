import React from "react";
import { AddressHistoryProps } from "../AddressHistory";

const AddressHistory: React.FC<AddressHistoryProps> = (props) => (
  <div className="address-history" />
);

export default AddressHistory;
