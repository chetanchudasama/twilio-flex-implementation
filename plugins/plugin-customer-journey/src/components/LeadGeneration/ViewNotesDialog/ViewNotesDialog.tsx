import React, {
  ChangeEvent,
  useMemo,
  useState,
  useRef,
  useEffect,
} from "react";
import { CustomInput, Shared, NoteIcon, secondary } from "@common/components";
import {
  Button,
  DialogActions,
  DialogContent,
  DialogTitle,
  Grid,
  IconButton,
} from "@material-ui/core";
import { v4 } from "uuid";
import { Manager } from "@twilio/flex-ui";
import CloseIcon from "@material-ui/icons/Close";

import { AddNoteRequestModel } from "../../../models/AddNoteRequestModel";
import { AppState } from "../../../states";
import { NoteModel } from "../../../models/NoteModel";
import { ViewNotesDialogStyles } from "./ViewNotesDialog.Styles";
import { ViewNotesDialogProps } from "./ViewNotesDialog.Props";
import { useApplicationService } from "../../../services/application.service";
import {
  CustomNotificationType,
  showErrorMessage,
  showMessage,
} from "../../../Notifications";

export const ViewNotesDialog: React.FC<ViewNotesDialogProps> = ({
  applicationId,
  customerName,
  open,
  recentNotes,
  dialogTitle,
  handleDialogClose,
}) => {
  const state: AppState = Manager.getInstance().store.getState();
  const token: string = useMemo(() => {
    return state.flex.session.ssoTokenPayload.token ?? "";
  }, [state.flex.session.ssoTokenPayload]);
  const applicationService = useApplicationService(token);

  const [notes, setNotes] = useState<string>("");
  const contentEndRef = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    // wait till note list bind in UI
    setTimeout(() => {
      if (contentEndRef && contentEndRef.current) {
        (contentEndRef.current as HTMLDivElement).scrollIntoView({
          behavior: "smooth",
        });
      }
    }, 0);
  }, []);

  const noteList = useMemo(() => {
    return recentNotes.sort((a: NoteModel, b: NoteModel) => {
      // eslint-disable-next-line no-nested-ternary
      return a.createdOn < b.createdOn ? -1 : a.createdOn > b.createdOn ? 1 : 0;
    });
  }, [recentNotes]);

  const handleNoteChange = (value: string) => {
    setNotes(value);
  };

  const onAddNotes = () => {
    const addNoteRequestModel = Object.assign(new AddNoteRequestModel(), {
      applicationId,
      notes,
    });

    applicationService
      .addNoteRequest(applicationId, addNoteRequestModel)
      .then(() => {
        handleDialogClose(notes);
        showMessage(
          CustomNotificationType.SuccessNotification,
          "Note detail saved successfully!"
        );
      })
      .catch((error) => {
        showErrorMessage(
          "Error submitting add note request, please try again!",
          "",
          true
        );
      });
  };

  return (
    <ViewNotesDialogStyles
      onClose={() => handleDialogClose()}
      open={open}
      maxWidth="sm"
      disableBackdropClick
      disableEscapeKeyDown
      scroll="paper"
    >
      <DialogTitle className="dialog-title">
        <div className="title-text">
          <NoteIcon color={secondary} />
          {`${dialogTitle}: ${customerName} (#${applicationId})`}
        </div>
        <IconButton className="close-icon" onClick={() => handleDialogClose()}>
          <CloseIcon />
        </IconButton>
      </DialogTitle>
      <DialogContent className="dialog-content">
        <Grid container className="notes-container">
          <Grid item xs={12}>
            {noteList.length === 0 && (
              <Grid item xs={12} className="note-unavailable">
                No note(s) available
              </Grid>
            )}
            {noteList.length > 0 &&
              noteList.map((note: NoteModel) => (
                <Grid container key={v4()} className="comment-container">
                  <Grid item xs={3} className="comment-date-time">
                    {Shared.getFormattedDate(note.createdOn, "DD/MM/YY HH:mm")}
                  </Grid>
                  <Grid item xs={9} className="comment-detail">
                    <Grid container>
                      <Grid item xs={12} className="comment">
                        {note.comment}
                      </Grid>
                      <Grid item xs={12} className="username">
                        {`Added by ${note.userName}`}
                      </Grid>
                    </Grid>
                  </Grid>
                </Grid>
              ))}
          </Grid>
        </Grid>
        <div ref={contentEndRef} className="content-end" />
      </DialogContent>
      <DialogActions className="dialog-action">
        <CustomInput
          label=""
          placeholder="Add a note"
          value={notes}
          onChange={(event: ChangeEvent<HTMLInputElement>) =>
            handleNoteChange(event.target.value)
          }
          rows={2}
          endAdornment={
            <Button
              onClick={onAddNotes}
              variant="contained"
              color="secondary"
              className={`add-btn ${!notes ? "disabled-btn" : ""}`}
              disabled={!notes}
            >
              Add
            </Button>
          }
          classes={{
            formControl: "custom-input",
          }}
        />
      </DialogActions>
    </ViewNotesDialogStyles>
  );
};
