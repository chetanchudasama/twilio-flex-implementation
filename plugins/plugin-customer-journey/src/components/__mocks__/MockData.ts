import { WorkerAttributes } from "@twilio/flex-ui";
import { namespace } from "../../states";

const workerAttributes: WorkerAttributes = {
  contact_uri: "client:test",
  email: "test@test.com",
  full_name: "test user",
  image_url: "img_url",
  roles: ["admin"],
  agentId: 1,
};

const conferenceSource = {
  source: {
    participants: [
      {
        callSid: "callSid",
        participantType: "customer",
        workerSid: "",
      },
      {
        callSid: "callSid",
        participantType: "worker",
        workerSid: "workerSid",
      },
    ],
  },
};

const conferenceState = new Map();
conferenceState.set("taskSid", conferenceSource);

export const flexManagerMock = jest.fn().mockReturnValue({
  workerClient: {
    source: {},
    sid: "workerSid",
    name: "test user",
    fullName: "test user",
    activityName: "Offline",
    attributes: workerAttributes,
    dateUpdated: new Date(),
    isAvailable: false,
    activityDuration: "",
    dateStatusChanged: new Date(1).toISOString(),
    activity: {
      available: false,
      name: "Offline",
      sid: "OfflineSid",
    },
  },
  store: {
    getState: jest.fn().mockReturnValue({
      flex: {
        worker: {
          attributes: workerAttributes,
        },
        session: {
          ssoTokenPayload: {
            token: "ssoToken",
          },
        },
        conferences: {
          states: conferenceState,
        },
      },
      [namespace]: {
        phoneNumberList: {
          phoneNumberList: [
            {
              friendlyName: "Sales",
              phoneNumber: "+1234567891",
              queueSid: "QUEUE_SID",
            },
          ],
        },
      },
    }),
  },
  strings: {
    DIALPADExternalTransferSearchInputLabel: "Search By Name or PhoneNumber",
    DIALPADExternalTransferHoverOver: "Add",
    DIALPADExternalTransferPopupHeader: "External Transfer",
  },
  serviceConfiguration: {
    outbound_call_flows: {
      default: {
        caller_id: "CALLER_ID",
      },
    },
  },
});
