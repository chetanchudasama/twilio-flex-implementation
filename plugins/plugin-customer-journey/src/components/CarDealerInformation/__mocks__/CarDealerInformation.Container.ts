import CarDealerInformation from "./CarDealerInformation";

export default CarDealerInformation;
