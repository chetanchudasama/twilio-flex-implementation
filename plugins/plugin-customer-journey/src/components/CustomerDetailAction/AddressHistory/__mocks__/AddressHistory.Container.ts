import AddressHistory from "./AddressHistory";

export default AddressHistory;
