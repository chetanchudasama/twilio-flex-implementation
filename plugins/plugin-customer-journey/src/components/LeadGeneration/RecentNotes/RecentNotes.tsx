import React, { useState, useMemo, useEffect } from "react";
import { Grid, Button, Divider } from "@material-ui/core";
import Typography from "@material-ui/core/Typography";
import { v4 } from "uuid";
import { Manager } from "@twilio/flex-ui";
import { Shared } from "@common/components";

import { AppState } from "../../../states";
import { RecentNotesStyles } from "./RecentNotes.Styles";
import { RecentNotesProps } from "./RecentNotes.Props";
import { NoteModel } from "../../../models/NoteModel";
import ViewNotesDialog from "../ViewNotesDialog/ViewNotesDialog.Container";

export const RecentNotes: React.FC<RecentNotesProps> = ({ recentNotes }) => {
  const state: AppState = Manager.getInstance().store.getState();
  const [openViewNotesDialog, setOpenViewNotesDialog] =
    useState<boolean>(false);
  const [allNoteList, setAllNoteList] = useState<NoteModel[]>([]);
  const [isViewNoteBtnClicked, setIsViewNoteBtnClicked] =
    useState<boolean>(false);

  const noteList = useMemo(() => {
    return allNoteList.slice(0, 3);
  }, [allNoteList]);

  const sortNoteList = (notes: NoteModel[]) => {
    return notes.sort((a: NoteModel, b: NoteModel) => {
      // eslint-disable-next-line no-nested-ternary
      return a.createdOn < b.createdOn ? 1 : a.createdOn > b.createdOn ? -1 : 0;
    });
  };

  useEffect(() => {
    let tempNoteList = [...allNoteList];
    const newNotes = recentNotes.filter(
      (n: NoteModel) => !allNoteList.includes(n)
    );
    tempNoteList = [...tempNoteList, ...newNotes];
    setAllNoteList(sortNoteList(tempNoteList));
    // useEffect need not to be called on allNoteList state change
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [recentNotes]);

  const handleViewNotesDialogClose = (notes?: string) => {
    if (notes) {
      // add note detail to recent note list
      const noteModel: NoteModel = Object.assign(new NoteModel(), {
        commentId: v4(),
        comment: notes,
        createdOn: new Date().toISOString(),
        userName: state.flex.worker.attributes?.full_name || "",
      });
      const tempNoteList = [...allNoteList];
      tempNoteList.push(noteModel);
      setAllNoteList(sortNoteList(tempNoteList));
    }
    setOpenViewNotesDialog(false);
    setIsViewNoteBtnClicked(false);
  };

  return (
    <>
      <RecentNotesStyles>
        <Grid item xs={12} className="note-title-container">
          <Grid
            container
            direction="row"
            justify="space-between"
            alignItems="center"
            spacing={8}
          >
            <Grid item>
              <Typography component="p" variant="caption" className="heading">
                Recent Notes
              </Typography>
            </Grid>
            <Grid item>
              <Button
                variant="text"
                className="note-btn view-note-btn"
                onClick={() => {
                  setIsViewNoteBtnClicked(true);
                  setOpenViewNotesDialog(true);
                }}
              >
                View all notes
              </Button>
              <Button
                variant="text"
                className="note-btn"
                onClick={() => {
                  setOpenViewNotesDialog(true);
                }}
              >
                Add note
              </Button>
            </Grid>
          </Grid>
        </Grid>
        <Grid item xs={12}>
          {recentNotes.length === 0 && (
            <Grid item xs={12} className="note-unavailable">
              No recent note(s) available
            </Grid>
          )}
          {noteList.length > 0 &&
            noteList.map((note: NoteModel, index: number) => (
              <Grid container key={v4()}>
                <Grid item xs={12} className="comment">
                  {note.comment}
                </Grid>
                <Grid item xs={12} className="comment-detail">
                  <span className="created-date">
                    {Shared.getFormattedDate(note.createdOn, "DD/MM/YY")}
                  </span>
                  <span>{note.userName}</span>
                </Grid>
                <Grid item xs={12}>
                  {noteList.length - 1 !== index && (
                    <Divider className="divider" />
                  )}
                </Grid>
              </Grid>
            ))}
        </Grid>
      </RecentNotesStyles>
      {openViewNotesDialog && (
        <ViewNotesDialog
          open={openViewNotesDialog}
          recentNotes={allNoteList}
          dialogTitle={isViewNoteBtnClicked ? "View Note" : "Add Note"}
          handleDialogClose={handleViewNotesDialogClose}
        />
      )}
    </>
  );
};
