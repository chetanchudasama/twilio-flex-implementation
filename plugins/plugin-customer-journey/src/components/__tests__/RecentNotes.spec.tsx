import React from "react";
import { create } from "react-test-renderer";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";

import { Grid } from "@material-ui/core";
import { Manager } from "@twilio/flex-ui";
import ViewNotesDialog from "../LeadGeneration/ViewNotesDialog/ViewNotesDialog.Container";
import { RecentNotesProps } from "../LeadGeneration/RecentNotes/RecentNotes.Props";
import { RecentNotes } from "../LeadGeneration/RecentNotes/RecentNotes";
import { flexManagerMock } from "../__mocks__/MockData";

Manager.getInstance = flexManagerMock;

jest.mock("../LeadGeneration/ViewNotesDialog/ViewNotesDialog.Container", () => {
  return {
    __esModule: true,
    default: jest.fn(({ children }) => {
      return <div className="view-notes-dialog">{children}</div>;
    }),
  };
});

const props: RecentNotesProps = {
  recentNotes: [
    {
      comment: "This is a test note",
      commentId: 1,
      createdOn: new Date(2021, 8, 8),
      department: "Development",
      userName: "User",
      isImportant: false,
    },
    {
      comment: "Test note",
      commentId: 11,
      createdOn: new Date(2021, 8, 4),
      department: "Development",
      userName: "User",
      isImportant: false,
    },
  ],
};

describe("RecentNotes", () => {
  it("renders correctly on load", () => {
    const wrapper = create(<RecentNotes {...props} />).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it("should render view all notes, add note buttons", () => {
    const wrapper = create(<RecentNotes {...props} />).root;

    expect(
      wrapper.findAllByProps({
        className: "note-btn",
      })
    ).toHaveLength(2);
  });

  it("should render recent notes list", () => {
    const wrapper = create(<RecentNotes {...props} />);

    const noteElements = wrapper.root.findAll(
      (el) => el.props.className === "comment" && el.type === Grid
    );
    const commentDetailElements = wrapper.root.findAll(
      (el) => el.props.className === "comment-detail" && el.type === Grid
    );

    expect(noteElements.length).toEqual(props.recentNotes.length);
    expect(commentDetailElements.length).toEqual(props.recentNotes.length);
  });

  it("should render no recent notes available message", () => {
    props.recentNotes = [];
    const wrapper = create(<RecentNotes {...props} />);

    const elements = wrapper.root.findAll(
      (el) => el.props.className === "note-unavailable" && el.type === Grid
    );

    expect(elements.length).toEqual(1);
  });

  it("should open ViewNotesDialog, when View all notes button is clicked", () => {
    const wrapper = mount(<RecentNotes {...props} />);

    const button = wrapper.find(".view-note-btn").last();
    button.simulate("click");

    expect(wrapper.find(ViewNotesDialog).exists()).toBeTruthy();
    expect(wrapper.find(ViewNotesDialog).props().dialogTitle).toBe("View Note");
  });

  it("should open ViewNotesDialog, when Add note button is clicked", () => {
    const wrapper = mount(<RecentNotes {...props} />);

    const button = wrapper.find(".note-btn").last();
    button.simulate("click");

    expect(wrapper.find(ViewNotesDialog).exists()).toBeTruthy();
    expect(wrapper.find(ViewNotesDialog).props().dialogTitle).toBe("Add Note");
  });

  it("should close ViewNotesDialog component, when handleDialogClose method is emitted", () => {
    const wrapper = mount(<RecentNotes {...props} />);

    const button = wrapper.find(".note-btn").last();
    button.simulate("click");

    expect(wrapper.find(ViewNotesDialog).exists()).toBeTruthy();

    act(() => {
      wrapper.find(ViewNotesDialog).props().handleDialogClose();
    });
    wrapper.update();

    expect(wrapper.find(ViewNotesDialog).exists()).toBe(false);
  });
});
