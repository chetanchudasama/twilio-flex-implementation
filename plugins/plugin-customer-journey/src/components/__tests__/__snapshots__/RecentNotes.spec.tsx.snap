// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentNotes renders correctly on load 1`] = `
<div
  className="css-1cmugtl"
>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 note-title-container"
  >
    <div
      className="MuiGrid-container-1 MuiGrid-spacing-xs-8-23 MuiGrid-align-items-xs-center-9 MuiGrid-justify-xs-space-between-20"
    >
      <div
        className="MuiGrid-item-2"
      >
        <p
          className="MuiTypography-root-98 MuiTypography-caption-108 heading"
        >
          Recent Notes
        </p>
      </div>
      <div
        className="MuiGrid-item-2"
      >
        <button
          className="MuiButtonBase-root-160 MuiButton-root-134 MuiButton-text-136 MuiButton-flat-139 note-btn view-note-btn"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-135"
          >
            View all notes
          </span>
          <span
            className="MuiTouchRipple-root-163"
          />
        </button>
        <button
          className="MuiButtonBase-root-160 MuiButton-root-134 MuiButton-text-136 MuiButton-flat-139 note-btn"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-135"
          >
            Add note
          </span>
          <span
            className="MuiTouchRipple-root-163"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-12-41"
  >
    <div
      className="MuiGrid-container-1"
    >
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 comment"
      >
        This is a test note
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 comment-detail"
      >
        <span
          className="created-date"
        >
          08/09/21
        </span>
        <span>
          User
        </span>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41"
      >
        <hr
          className="MuiDivider-root-170 divider"
        />
      </div>
    </div>
    <div
      className="MuiGrid-container-1"
    >
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 comment"
      >
        Test note
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 comment-detail"
      >
        <span
          className="created-date"
        >
          04/09/21
        </span>
        <span>
          User
        </span>
      </div>
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41"
      />
    </div>
  </div>
</div>
`;
